### DISCUSSION TITLE
NatsConnection restart writer first , then update `currentServer` ?

### DISCUSSION THREAD

**Discussion by pg-yang** [nats-io/nats.java#711] [Q&A]
Hello , I 'm writing a  apm plugin , It use java agent  technology .  And I try to  intercept `NatsConnectionWriter#sendMessageBatch` method  .

To get current correct server URL , I call `NatsConnection#getConnectedUrl` through `NatsConnectionWriter#connection` field
But I found , when `tryToConnect` , The library restart `NatsConnectionWriter` first  , then update  server URL. like this:

```
 void tryToConnect(String serverURI, long now) {
     //.....
     this.dataPortFuture.complete(this.dataPort);
     this.writer.start(this.dataPortFuture);
      // 
     this.currentServer = serverURI;
 }
```
Maybe it's more reasonable  that  update currentServer first , then restart writer thread ? 
Any suggestion for get real server uri when call `NatsConnectionWriter#sendMessageBatch` ?  if I get the url from `NatsConnectionWriter#sendMessageBatch` method argument (DataPort dataPort) , I have to use java reflect , Because SocketDataPort filed is private

**Comment by pg-yang** [comment]
@scottf  Could you help me ?

**Reply by scottf** [reply]
What if the connection fails? I wonder if that was the intention of the way things are ordered.
